upf/upf_pfcp_api.c

```diff
@@ -657,28 +657,60 @@ handle_f_teid (upf_session_t * sx, upf_main_t * gtm, pfcp_pdi_t * pdi,
 
 }
 
+static void
+expunge_conflicting_session (upf_session_t * sx_old,
+			     upf_session_t * sx_new,
+			     pfcp_ue_ip_address_t * ue_addr)
+{
+  int r;
+  upf_main_t *gtm = &upf_main;
+
+  clib_warning ("UE IP conflict: deleting session CP SEID=0x%016" PRIx64
+		" %U " "due to attempt to create CP SEID=0x%0x16" PRIx64,
+		sx_old->cp_seid, format_ue_ip_address, ue_addr,
+		sx_new->cp_seid);
+
+  upf_pfcp_server_deferred_free_msgs_by_sidx (sx_old - gtm->sessions);
+  upf_pfcp_session_up_deletion_report (sx_old);
+
+  if ((r = pfcp_disable_session (sx_old)) != 0)
+    clib_warning ("PFCP Session %" PRIu64 " could no be disabled\n",
+		  sx_old->cp_seid);
+  else
+    pfcp_free_session (sx_old);
+}
+
 static int
-validate_ue_ip (upf_session_t * sx, upf_nwi_t * nwi,
-		pfcp_ue_ip_address_t * ue_addr)
+resolve_ue_ip_conflicts (upf_session_t * sx, upf_nwi_t * nwi,
+			 pfcp_ue_ip_address_t * ue_addr)
 {
   upf_main_t *gtm = &upf_main;
   const dpo_id_t *dpo;
+  int r = 0;
 
   if (ue_addr->flags & IE_UE_IP_ADDRESS_V4)
     {
       dpo = upf_get_session_dpo_ip4 (nwi, &ue_addr->ip4);
       if (dpo && dpo->dpoi_index != sx - gtm->sessions)
-	return -1;
+	{
```

Comment:
a real problem only exists if the same CP node tries to use the same UE IP twice. If the UE IP was allocated by a different node, that other node would still be aware of the assignment and there would be no further problems.

The function should therefore check for the node_id and only remove the other session if the node_ids match.

Reply:
Fixed. If node id is different, the new Session Establishment Request will still fail, but the old session will not be dropped.